proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+Cryptography in Haskell is also fragile and outdated, having seen considerable flux / churn over the years as various libraries have been developed and then deprecated or abandoned in favor of newer ones. This has placed the long-term stability of important libraries such as `tls` and `x509` at risk of falling behind as compilers are upgraded, old standards are updated, and new standards are implemented.
+
+There are several reasons for this:
+
+1) Cryptography involves a lot of stateful low-level bit twiddling and random IO, something Haskell is not known for being good at. Classes like `Bits` and `FiniteBits` classes are critical for cryptography, but are terribly awkward to implement (like `Num`). Also, `ByteString` does not implement them for nuanced reasons.
+
+2) Cryptography has a high knowledge and technical skill floor. Cryptography is more than just the usual primitives of ciphers, public and private keys, hashes, and MACs, and the scope of knowledge can grow quickly as primitives are combined into more advanced algorithms.
+
+3) The notion that functional programming is "bad at cryptography" (ie, underperformant and has insurmountable challenges) is a self-fulfilling prophecy, and will remain so unless we invest in enough effort to disprove it.
+
+4) A fair lot of cryptographic projects in recent years has been specifically towards *cryptocurrencies*, rather than *cryptography in general*, and their results are at large fairly useless if they cannot be used outside of their ecosystems.
+
+All of these reasons have contributed to the stagnant state of Haskell cryptography, and have make it an unfavorable environment for developers of cryptosystems and distributed systems. However, none of these are actual blockers, and each of these things can be corrected with effort. There is much that can be done to improve the Haskell cryptography ecosystem, ranging from better low-level support, to higher-level abstractions:
+
+- Up-to-date algorithms
+- Hierarchy of typeclasses
+- Monadic interfaces
+- Safety abstractions
+- Type-level composition
+- Advanced data structure (eg, merkle trees)
+- Use of Generic for crypto-serialization
+
+As a result, there is much to be gained from investing in foundational Haskell cryptography. Haskell is well-suited to providing many aspects of information security through use of monadic interfaces and the type system, if the lower-level challenges of strictness and performant primitives can be solved. The community needs a cryptography library that is easily maintained, with minimal dependencies / baggage, and with documentation, tutorials, and unit tests to ease development.
+
+A first step would be developing bindings to a suitably-licensed open-source C or C++ library containing modern and up-to-date cryptography algorithms, including post-quantum cryptography schemes.
```

Comment:
While  I quibble with your steps, I think the conclusion is correct. Here's what I think is a more convincing argument:
1. Resistance to side channel attacks requires very low-level control over an implementation, but part of the point of Haskell is getting away from micromanaging the computer.
2. Trustworthy crypto libraries are extremely expensive to develop, but bindings are much cheaper
3. Haskell's C FFI can connect well to C implementations
4. The crypto community knows C, they don't know Haskell, so there's more resources to get the nonfunctional requirements right over there, and mature libraries exist.

Reply:
This is a much more succinct way of expressing some of the things that I have been attempting to say - I will revise them with this in mind.